chia/full_node/full_node.py

```diff
@@ -1068,8 +1079,34 @@ async def validate_block_batches(
                 peer, blocks = res
                 start_height = blocks[0].height
                 end_height = blocks[-1].height
+
+                # in case we're validating a reorg fork (i.e. not extending the
+                # main chain), we need to record the coin set from that fork in
+                # fork_info. Otherwise validation is very expensive, especially
+                # for deep reorgs
+                peak: Optional[BlockRecord]
+                if fork_info is None:
```

Comment:
this is a long running task that validates blocks, pipelined from the other task that requests them from the peer.
Until this point, we don't have any actual blocks to validate, so we don't know whether we're extending the existing longest chain or starting a reorg sync. This is important because we definitely don't want to create a `ForkInfo` object in the former case, as it would record additions and removals and just use up memory for no reason.